src/ppm.rs

```diff
@@ -0,0 +1,473 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements core functionality of the PPM protocol for Prio. It consists of the
+//! following functions:
+//!
+//! * `prio_upload` is run by the client to generate the input and proof shares.
+//!
+//! * `prio_verify_start` is run by each aggregator upon receiving its input and proof share. The
+//! output is broadcast to the other aggregators.
+//!
+//! * `prio_verify_finish` is run by each aggregator to decide if the input was valid.
+//!
+//! In the special case of two aggregators (one leader, one helper), these calls map to HTTP
+//! requests as follows (assuming they agree on the query randomness in advance):
+//!
+//! ```none
+//! Client                 Leader                    Helper
+//!   | prio_upload()        |                         |
+//!   |                      |                         |
+//!   |    upload request    |                         |
+//!   |--------------------->| prio_verify_start()     |
+//!   |                      |                         |
+//!   |                      |    aggregate request    |
+//!   |                      |------------------------>| prio_verify_start()
+//!   |                      |                         |
+//!   |                      |                         |
+//!   |                      |<------------------------|
+//!   |                      | prio_verify_finish()    | prio_verify_finish()
+//!   v                      v                         v
+//!                        output share              output share
+//! ```
+//!
+//! NOTE: This protocol implemented here is a prototype and has not undergone security analysis.
+//! Use at your own risk.
+
+use std::convert::TryFrom;
+
+use getrandom::getrandom;
+
+use crate::field::{rand, FieldElement};
+use crate::pcp::{decide, prove, query, PcpError, Proof, Value, Verifier};
+use crate::prng::{Prng, PrngError, SEED_LENGTH};
+
+/// Errors emitted by this module.
+#[derive(Debug, thiserror::Error)]
+pub enum PpmError {
+    /// An error occurred.
+    #[error("ppm error: {0}")]
+    Ppm(String),
+
+    /// The distributed input was deemed invalid.
+    #[error("ppm error: invalid distributed input: {0}")]
+    Validity(&'static str),
+
+    /// PCP error.
+    #[error("ppm error: pcp error: {0}")]
+    Pcp(#[from] PcpError),
+
+    /// PRNG error.
+    #[error("ppm error: prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// Calling get_random() returned an error.
+    #[error("ppm error: getrandom error: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// A share of an input or proof for Prio.
+pub enum PrioShare<F: FieldElement> {
+    /// An uncompressed share, typically sent to the leader.
+    Leader(Vec<F>),
+
+    /// A compressed share, typically sent to the helper. The first parameter is a seed for a
+    /// pseudorandom generator
+    Helper(Vec<u8>, usize),
```

Comment:
I'm happy to defer discussion of `SEED_LENGTH` until the `Prng` API solidifies, but I think this would still be better:
```suggestion
    Helper {
        seed: Vec<u8>,
        uncompressed_length: usize,
    },
```